src/backend/distributed/metadata/metadata_utility.c

```diff
@@ -2357,3 +2379,1758 @@ IsForeignTable(Oid relationId)
+	{
+		return BACKGROUND_TASK_STATUS_UNSCHEDULED;
+	}
+	else if (enumOid == CitusTaskStatusBlockedId())
+	{
+		return BACKGROUND_TASK_STATUS_BLOCKED;
+	}
+	else if (enumOid == CitusTaskStatusCancelledId())
+	{
+		return BACKGROUND_TASK_STATUS_CANCELLED;
+	}
+	else if (enumOid == CitusTaskStatusCancellingId())
+	{
+		return BACKGROUND_TASK_STATUS_CANCELLING;
+	}
+	ereport(ERROR, (errmsg("unknown enum value for citus_task_status")));
+}
+
+
+/*
+ * IsBackgroundJobStatusTerminal is a predicate returning if the BackgroundJobStatus
+ * passed is a terminal state of the Background Job state machine.
+ *
+ * For a Job to be in it's terminal state, all tasks from that job should also be in their
+ * terminal state.
+ */
+bool
+IsBackgroundJobStatusTerminal(BackgroundJobStatus status)
+{
+	switch (status)
+	{
+		case BACKGROUND_JOB_STATUS_CANCELLED:
+		case BACKGROUND_JOB_STATUS_FAILED:
+		case BACKGROUND_JOB_STATUS_FINISHED:
+		{
+			return true;
+		}
+
+		case BACKGROUND_JOB_STATUS_CANCELLING:
+		case BACKGROUND_JOB_STATUS_FAILING:
+		case BACKGROUND_JOB_STATUS_RUNNING:
+		case BACKGROUND_JOB_STATUS_SCHEDULED:
+		{
+			return false;
+		}
+
+			/* no default to make sure we explicitly add every state here */
+	}
+	elog(ERROR, "unknown BackgroundJobStatus");
+}
+
+
+/*
+ * IsBackgroundTaskStatusTerminal is a predicate returning if the BackgroundTaskStatus
+ * passed is a terminal state of the Background Task state machine.
+ */
+bool
+IsBackgroundTaskStatusTerminal(BackgroundTaskStatus status)
+{
+	switch (status)
+	{
+		case BACKGROUND_TASK_STATUS_CANCELLED:
+		case BACKGROUND_TASK_STATUS_DONE:
+		case BACKGROUND_TASK_STATUS_ERROR:
+		case BACKGROUND_TASK_STATUS_UNSCHEDULED:
+		{
+			return true;
+		}
+
+		case BACKGROUND_TASK_STATUS_BLOCKED:
+		case BACKGROUND_TASK_STATUS_CANCELLING:
+		case BACKGROUND_TASK_STATUS_RUNNABLE:
+		case BACKGROUND_TASK_STATUS_RUNNING:
+		{
+			return false;
+		}
+
+			/* no default to make sure we explicitly add every state here */
+	}
+	elog(ERROR, "unknown BackgroundTaskStatus");
+}
+
+
+/*
+ * BackgroundJobStatusOid returns the Oid corresponding to SQL enum value corresponding to
+ * the BackgroundJobStatus.
+ */
+Oid
+BackgroundJobStatusOid(BackgroundJobStatus status)
+{
+	switch (status)
+	{
+		case BACKGROUND_JOB_STATUS_SCHEDULED:
+		{
+			return CitusJobStatusScheduledId();
+		}
+
+		case BACKGROUND_JOB_STATUS_RUNNING:
+		{
+			return CitusJobStatusRunningId();
+		}
+
+		case BACKGROUND_JOB_STATUS_CANCELLING:
+		{
+			return CitusJobStatusCancellingId();
+		}
+
+		case BACKGROUND_JOB_STATUS_FINISHED:
+		{
+			return CitusJobStatusFinishedId();
+		}
+
+		case BACKGROUND_JOB_STATUS_CANCELLED:
+		{
+			return CitusJobStatusCancelledId();
+		}
+
+		case BACKGROUND_JOB_STATUS_FAILING:
+		{
+			return CitusJobStatusFailingId();
+		}
+
+		case BACKGROUND_JOB_STATUS_FAILED:
+		{
+			return CitusJobStatusFailedId();
+		}
+	}
+
+	elog(ERROR, "unknown BackgroundJobStatus");
+	return InvalidOid;
```

Comment:
fwiw, we have various other cases where we ERROR on the last line without a return and without compilers complaining. Might be a matter of using ereport instead of elog.